Convert the rest of interop for Cursor

JeremyKuhne opened the pull request:
## Proposed changes

- Make all of Cursor Interop clean.
- Update other classes with related imports to use static Interop (which we're doing in the designer)
- Clean style in DeviceContext and WindowsGraphics
- Introduce internal IHandle to make GC.KeepAlive wraps easier

## Customer Impact

- Better perf for IPicture related code paths
- More robust for said code paths

## Regression? 

- No

## Risk

- Low

## Test methodology

- Ensured areas had test coverage and ran tests


Note that I wanted to get this up to demonstrate some patterns, notably:

1. Using `IHandle` to start providing keepalive overloads
2. Using static for `Interop` (as we're doing in the designer repo now)
3. Union handling (see `PICTDESC`)
4. Generic pattern for `GetObject`
5. Ref disposable pattern on native structs (see `ICONINFO`)

Whether or not we take it for 3.0 I wanted these examples visible.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1604)

RussKie commented on src/Common/src/Interop/User32/Interop.CopyImage.cs:
```diff
@@ -0,0 +1,50 @@
+            return result;
+        }
+
+
```
[nit] extra line

RussKie commented on src/Common/src/Interop/Ole32/Interop.PICTYPE.cs:
```diff
@@ -0,0 +1,23 @@
+        /// </summary>
+        public enum PICTYPE : uint
+        {
+            UNINITIALIZED = unchecked((uint)(-1)),
```
Could you please explain this?
The original was `int`, the signature for `IPicture::get_Type` uses `SHORT`. Why are we not using `short`?

weltkante replied:
* [PICTDESC](https://docs.microsoft.com/en-us/windows/win32/api/olectl/ns-olectl-pictdesc) uses `UINT`
* `PICTYPE_UNINITIALIZED` is defined as `#define PICTYPE_UNINITIALIZED (-1)` (without typecast to unsigned!)
* [IPicture::get_Type](https://docs.microsoft.com/de-DE/windows/win32/api/ocidl/nf-ocidl-ipicture-get_type) uses `SHORT`

Considering how `PICTYPE_UNINITIALIZED` is defined `int` or `short` would be the correct choice, regardless of how `PICTDESC` is defined. I believe the Win32 API is in error here because sign extension from `SHORT` to `UINT` will lead to the wrong constant being used. Using `-1` as a constant with different widths requires sign extension so all variables should be signed for it to work correctly and minimize potential sources of errors.

JeremyKuhne replied:
`PICTDESC` uses UINT. It allowed making it blittable. While `UNITIALIZED` is `(-1)`, it's super common to have unchecked casts like this to get `0xFFFF` etc. I prefer it to be unsigned as signed values have no meaning in this context. I've repeatedly seen code do validations on sign of values when it doesn't have anything to do with validity for the native definition (think widths, height, length, etc.). As such I lean towards unsigned where there is some ambiguity unless there is explicit meaning to unsigned values in general.

weltkante replied:
> it's super common to have unchecked casts like this to get 0xFFFF etc.

I worry about the reverse, that `SHORT` -1 (0xFFFF) may not always be sign-extended to `UINT` 0xFFFF_FFFF

Will a `SHORT` get correctly sign-extended to `UINT` when assigning the result of get_Type to a `PICTDESC`? I don't know the sign-extending rules for unsigned types of different widths that well, using a signed value seems more safe.

JeremyKuhne replied:
I need it to be the same size as PICTDESC, I'll go ahead and make this one `int`.

RussKie commented on src/Common/src/Interop/User32/Interop.DestroyCursor.cs:
```diff
@@ -0,0 +1,16 @@
+    internal static partial class User32
+    {
+        [DllImport(Libraries.User32, ExactSpelling = true)]
+
```
Blank line

RussKie commented on src/Common/src/Interop/User32/Interop.ShowCursor.cs:
```diff
@@ -0,0 +1,15 @@
+    internal static partial class User32
+    {
+        [DllImport(Libraries.User32, ExactSpelling = true)]
+
```
Blank line

RussKie commented on src/Common/src/UnsafeNativeMethods.cs:
```diff
@@ -366,8 +349,7 @@ public static IntPtr ChildWindowFromPointEx(HandleRef hwndParent, int x, int y,
 
-        [DllImport(ExternDll.User32, ExactSpelling = true, CharSet = CharSet.Auto)]
-        public static extern bool GetCursorPos(out Point pt);
+
```
[nit] blank line

JeremyKuhne commented:
Rebased and addressed feedback.

weltkante commented on src/Common/src/Interop/Ole32/Interop.PICTYPE.cs:
```diff
@@ -0,0 +1,23 @@
+        ///  Picture type used with <see cref="IPicture"/> and
+        ///  <see href="https://docs.microsoft.com/en-us/windows/win32/com/pictype-constants"/>
+        /// </summary>
+        public enum PICTYPE : int
```
Do we want to add a comment why this is `int`?
```suggestion
        /// <remarks>
        ///  This type is used with different field sizes, it is signed to prevent
        ///  sign-extension errors when going from smaller to larger field sizes.
        /// </remark>
        public enum PICTYPE : int
```

JeremyKuhne commented:
I had the interfaces wrong and didn't notice until I tried using IPictureDisp. It has always been incorrect.

`OLE_HANDLE` is a uint, no matter what the platform. The dispatch invocation throws as it sees that the native type is uint, which doesn't match IntPtr. Invoking through the normal IUnknown RCW doesn't have any type info to check so it works fine due to the luck of calling conventions matching up.

The IPictureDisp version never worked on 64 bit. Now it does. :)

JeremyKuhne commented:
I added converter tests for AxHosts that exercise these interfaces fairly directly. I also moved a couple monitor methods to the interop folder when I was playing around with creating the tests.

JeremyKuhne commented:
I added tests for the rest of the IPicture usages (in the COM converter) and fixed a few issues with said converter. I introduced a helper to allow testing when the debug code asserts and moved himetric conversion to the common test library.

RussKie commented on src/Common/tests/InternalUtilitiesForTests/InternalUtilitiesForTests.csproj:
```diff
@@ -13,4 +12,8 @@
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\System.Windows.Forms\src\System.Windows.Forms.csproj" />
+  </ItemGroup>
```
I'm not particularly happy with adding this dependency here. `GdiHelper` is System.Windows.Forms specific, so it doesn't feel correct to add it to a shared project.
For now I'd move `GdiHelper` to the interop test project, but later, if becomes necessary, I'd create a System.Windows.Forms.TestCommon.

JeremyKuhne replied:
It isn't `System.Windows.Forms` specific. Any sort of test that needs to get logical pixels, convert units, etc. can utilize this. I can link in the source files instead if that works.

RussKie commented on src/Common/tests/InternalUtilitiesForTests/src/NoAssertListener.cs:
```diff
@@ -0,0 +1,100 @@
+    /// <summary>
+    ///  Use (within a using) to eat asserts.
+    /// </summary>
+    public sealed class NoAssertListener : TraceListener
```
We already have `TraceListenerlessContext` that ignores asserts
https://github.com/dotnet/winforms/blob/260410e7c97325ca434baf75e1bae0d2882a1afa/src/Common/tests/TraceListenerlessContext.cs

JeremyKuhne replied:
Thanks for the pointer, I'll replace that one as it isn't correct: multi-threaded use will cause you both get false test failures and miss valid ones (listeners are process wide). I presume it was just trying to avoid debug fails and not truly trying to remove all listeners. If it is actually trying to kill all listeners I'll fix it and add this one.

JeremyKuhne replied:
The other problem with removing all listeners is that you prevent any other listener from doing it's job across the whole process. Better to be targeted to exactly where we want the change.

JeremyKuhne replied:
That said, my implementation has holes. Fixing them. :)

RussKie replied:
Thanks, yep the original intent was to swallow asserts that were breaking tests under specifically crafted conditions.
Feel free to tweak/fix/nuked it as necessary.

JeremyKuhne commented:
Rebased.
